includes/class-llms-order-generator.php

```diff
@@ -0,0 +1,733 @@
+
+		$data = get_class_vars( __CLASS__ );
+		foreach ( $data as $key => &$val ) {
+			$val = $this->{$key};
+		}
+
+		return new WP_Error( $code, $message, array_merge( $data, $extra_data ) );
+
+	}
+
+	/**
+	 * Attempts to locate a user ID.
+	 *
+	 * Uses the logged in users information and falls back to a lookup by email address if available.
+	 *
+	 * @since [version]
+	 *
+	 * @param string|null $email An email address, if available.
+	 * @return null|integer Returns the WP_User ID or null if not found.
+	 */
+	private function find_user_id( $email = null ) {
+
+		if ( is_user_logged_in() ) {
+			return get_current_user_id();
+		}
+
+		if ( $email ) {
+			$user = get_user_by( 'email', $email );
+			return $user ? $user->ID : null;
+		}
+
+		return null;
+
+	}
+
+	/**
+	 * Generates an order.
+	 *
+	 * Uses data submitted during class construction and performs all necessary
+	 * validations. If validations pass, creates the order.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $user_action The user action, accepts `LLMS_Order_Generator::UA_COMMIT` or `LLMS_Order_Generator::UA_VALIDATE`.
+	 * @return WP_Error|LLMS_Order
+	 */
+	public function generate( $user_action = self::UA_COMMIT ) {
+
+		$validate = $this->validate();
+		if ( is_wp_error( $validate ) ) {
+			return $validate;
+		}
+
+		if ( self::UA_COMMIT === $user_action ) {
+			$user = $this->commit_user();
+			if ( is_wp_error( $user ) ) {
+				return $user;
+			}
+		}
+
+		return $this->create();
+
+	}
+
+	/**
+	 * Retrieves the coupon object for the order.
+	 *
+	 * @since [version]
+	 *
+	 * @return LLMS_Coupon|null
+	 */
+	public function get_coupon() {
+		return $this->coupon;
+	}
+
+	/**
+	 * Retrieves the payment gateway instance for the order.
+	 *
+	 * @since [version]
+	 *
+	 * @return LLMS_Payment_Gateway|null
+	 */
+	public function get_gateway() {
+		return $this->gateway;
+	}
+
+	/**
+	 * Retrieves the order id to use for the order.
+	 *
+	 * Attempts to locate an existing pending order by order key if it was submitted,
+	 * otherwise returns `new` which denotes a new order should be created.
+	 *
+	 * @since [version]
+	 *
+	 * @return integer|string
+	 */
+	protected function get_order_id() {
+
+		$order_id = null;
+		$key      = $this->data['llms_order_key'] ?? null;
+		$email    = $this->data['email_address'] ?? null;
+		$plan_id  = $this->data['llms_plan_id'] ?? null;
+
+		// Try to lookup using the order key if it was supplied.
+		if ( $key ) {
+			$order_id = llms_get_order_by_key( $key, 'id' );
+		}
+
+		// Try to lookup by user ID.
+		if ( ! $order_id ) {
+
+			$user_id  = $this->find_user_id( $email );
+			$order_id = $user_id ? llms_locate_order_for_user_and_plan( $user_id, $plan_id ) : $order_id;
+
+		}
+
+		// Lookup by email address.
+		if ( ! $order_id && $email ) {
+			$order_id = llms_locate_order_for_email_and_plan( $email, $plan_id );
+		}
+
+		return $order_id ? $order_id : 'new';
```

Comment:
```suggestion
		return $order_id ?: 'new';
```
[Ternary Operator](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.ternary)

Reply:
I'm sure we've talked about this already but if we haven't we're *not* using short ternary because WordPress doesn't: https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/

Reply:
Sorry. I forgot.
https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#ternary-operator

`??` null coalescing operator - good
`?:` ternary operator - bad